mysqld_exporter.go

```diff
@@ -172,6 +174,66 @@ func init() {
 	prometheus.MustRegister(version.NewCollector("mysqld_exporter"))
 }
 
+func filter(filters map[string]bool, name string, flag bool) bool {
+	if len(filters) > 0 {
+		return flag && filters[name]
+	}
+	return flag
+}
+
+func handler(w http.ResponseWriter, r *http.Request) {
+	var filters map[string]bool
+	params := r.URL.Query()["collect[]"]
+	log.Debugln("collect query:", params)
+
+	if len(params) > 0 {
+		filters = make(map[string]bool)
+		for _, param := range params {
+			filters[param] = true
+		}
+	}
+
+	collect := collector.Collect{
+		SlowLogFilter:        *slowLogFilter,
+		Processlist:          filter(filters, "info_schema.processlist", *collectProcesslist),
+		TableSchema:          filter(filters, "info_schema.tables", *collectTableSchema),
+		InnodbTablespaces:    filter(filters, "info_schema.innodb_tablespaces", *collectInnodbTablespaces),
+		InnodbMetrics:        filter(filters, "info_schema.innodb_metrics", *collectInnodbMetrics),
+		GlobalStatus:         filter(filters, "global_status", *collectGlobalStatus),
+		GlobalVariables:      filter(filters, "global_variables", *collectGlobalVariables),
+		SlaveStatus:          filter(filters, "slave_status", *collectSlaveStatus),
+		AutoIncrementColumns: filter(filters, "auto_increment.columns", *collectAutoIncrementColumns),
+		BinlogSize:           filter(filters, "binlog_size", *collectBinlogSize),
+		PerfTableIOWaits:     filter(filters, "perf_schema.tableiowaits", *collectPerfTableIOWaits),
+		PerfIndexIOWaits:     filter(filters, "perf_schema.indexiowaits", *collectPerfIndexIOWaits),
+		PerfTableLockWaits:   filter(filters, "perf_schema.tablelocks", *collectPerfTableLockWaits),
+		PerfEventsStatements: filter(filters, "perf_schema.eventsstatements", *collectPerfEventsStatements),
+		PerfEventsWaits:      filter(filters, "perf_schema.eventswaits", *collectPerfEventsWaits),
+		PerfFileEvents:       filter(filters, "perf_schema.file_events", *collectPerfFileEvents),
+		PerfFileInstances:    filter(filters, "perf_schema.file_instances", *collectPerfFileInstances),
+		UserStat:             filter(filters, "info_schema.userstats", *collectUserStat),
+		ClientStat:           filter(filters, "info_schema.clientstats", *collectClientStat),
+		TableStat:            filter(filters, "info_schema.tablestats", *collectTableStat),
+		QueryResponseTime:    filter(filters, "info_schema.query_response_time", *collectQueryResponseTime),
+		EngineTokudbStatus:   filter(filters, "engine_tokudb_status", *collectEngineTokudbStatus),
+		EngineInnodbStatus:   filter(filters, "engine_innodb_status", *collectEngineInnodbStatus),
+		Heartbeat:            filter(filters, "heartbeat", *collectHeartbeat),
+		HeartbeatDatabase:    *collectHeartbeatDatabase,
+		HeartbeatTable:       *collectHeartbeatTable,
+	}
+
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(collector.New(dsn, collect))
```

Comment:
@SuperQ @siavashs I think this broke `scrapes_total` and `scrape_errors_total` metrics. Previously `collector.New()` was created once - in `main` func. Now it's created on every scrape and so is `scrapes_total` and `scrape_errors_total` leading to constant values as every time `scrapes_total` is reset to 0, so we get only 1 scrape. It's not `total` anymore.
https://github.com/prometheus/mysqld_exporter/blob/master/collector/exporter.go#L107-L118

Reply:
I can confirm this bug, I think we can fix it by having a separate collector for exporter metrics.

Reply:
Sounds correct to me.  Thanks for the report.  I'll make a separate issue for this.

Reply:
For the reference: https://github.com/prometheus/mysqld_exporter/issues/269

Reply:
@SuperQ @siavashs Actually, there is one more confusing thing. Looks like now this metric is constant `2`, and even after fixing it, this will be increased every http call by `2` because each call ask for metrics and their description, and `Describe` again increases `scrapes_total` metric. Should `scrapes_total` be increased also when `Describe` is called? This way each http call would increase this metric only by `1`. I'm working right now on fix for this so let me please know ;)